### PR TITLE
Allow synchronous router migration

### PR DESCRIPTION
**Pull request by rhafer**
The new --wait-for-ports causes neutron-ha-tool to wait for all
not-distributed router-ports to be back in the ACTIVE state on the
target agent when migrating a router. This can be used to avoid long
work queues to pile up in the neutron agents when migrating many routers
in one go. And by that reduce the downtime per migrated router.

### CODE REVIEW

File: files/default/neutron-ha-tool.py [hunk 1]
```diff
@@ -81,6 +81,9 @@ def parse_args():
                     help='Replicate DHCP configuration to all agents')
     ap.add_argument('--now', action='store_true', default=False,
                     help='Migrate Routers immediately without a delay.')
+    ap.add_argument('--wait-for-ports', action='store_true', default=False,
```
Comment (aspiers) [inline files/default/neutron-ha-tool.py]: Better to default to `True`?

File: files/default/neutron-ha-tool.py [hunk 2]
```diff
@@ -219,18 +222,18 @@ def run(args):
     elif args.l3_agent_migrate:
         LOG.info("Performing L3 Agent Migration for Offline L3 Agents")
         errors = retry_with_backoff(l3_agent_migrate, args)(
-            qclient, args.noop, args.now)
+            qclient, args.noop, args.now, args.wait_for_ports)
```
Comment (aspiers) [inline files/default/neutron-ha-tool.py]: Might be better to just pass `args` for simplicity, otherwise there's a lot of argument passing.  Not 100% sure though.

File: files/default/neutron-ha-tool.py [hunk 3]
```diff
@@ -585,8 +589,54 @@ def migrate_router(qclient, router_id, agent, target):
...
+    :param target_host: The host on which the ports should be ACTIVE
+                        (reflected by the binding:host_id attribute of
+                        a port)
+    :param maxtries: Maximum number of times the ports' status is polled,
+                     an expection is raised there are still ports with a
+                     non ACTIVE status after that.
+    """
+
+    # FIXME: Don't check the "network:router_interface_distributed" ports
+    # as those are not hosted on the network node
```
Comment (aspiers) [inline files/default/neutron-ha-tool.py]: So this is currently broken for DVR?  That would probably be a blocker for merging :-/

File: files/default/neutron-ha-tool.py [hunk 4]
```diff
@@ -585,8 +589,54 @@ def migrate_router(qclient, router_id, agent, target):
...
+    router_port_list = qclient.list_ports(device_id=router_id, fields='id')
```
Comment (aspiers) [inline files/default/neutron-ha-tool.py]: Out of curiosity how did you figure out this API?  The only doc I could find was http://docs.openstack.org/user-guide/sdk-neutron-apis.html which doesn't detail the parameters.

File: files/default/neutron-ha-tool.py [hunk 5]
```diff
@@ -585,8 +589,54 @@ def migrate_router(qclient, router_id, agent, target):
...
+    port_ids = [r['id'] for r in router_port_list['ports']]
+
+    LOG.info("Wait for the ports of router_id=%s to be ACTIVE" % router_id)
+    while (maxtries > 0) and (len(port_ids) > 0):
```
Comment (aspiers) [inline files/default/neutron-ha-tool.py]: Isn't this a bit overkill on the parentheses? :-)  An ignorable nitpick though.

File: files/default/neutron-ha-tool.py [hunk 6]
```diff
@@ -585,8 +589,54 @@ def migrate_router(qclient, router_id, agent, target):
...
+        remaining = []
+        for port_id in port_ids:
+            port = qclient.show_port(port_id)['port']
```
Comment (aspiers) [inline files/default/neutron-ha-tool.py]: Is it possible to retrieve this in a single API call, rather than one per port?  If so, that could reduce overall time taken for migrating all routers.

File: files/default/neutron-ha-tool.py [hunk 7]
```diff
@@ -585,8 +589,54 @@ def migrate_router(qclient, router_id, agent, target):
...
+            else:
+                remaining.append(port_id)
+
+        if len(remaining) > 0:
+            time.sleep(1)
+            maxtries-=1
+        port_ids = remaining
+
+    if len(port_ids):
+        raise RuntimeError("Some ports are not ACTIVE on router_id=%s, ports=[%s]" %
```
Comment (aspiers) [inline files/default/neutron-ha-tool.py]: I don't think you need to repeat the word `ports` here:

    raise RuntimeError("Some ports are not ACTIVE on router_id=%s: [%s]" %

Small nitpick but it's a pain when log lines are long enough to scroll off the right-hand edge of the terminal ...

File: files/default/neutron-ha-tool.py [hunk 8]
```diff
@@ -585,8 +589,54 @@ def migrate_router(qclient, router_id, agent, target):
...
+
+    LOG.info("Wait for the ports of router_id=%s to be ACTIVE" % router_id)
+    while (maxtries > 0) and (len(port_ids) > 0):
+        remaining = []
+        for port_id in port_ids:
+            port = qclient.show_port(port_id)['port']
+            LOG.debug("Router port=%s on hostid=%s in status=%s" %
+                      (port['id'], port['binding:host_id'],
+                       port['status']))
+            if ((port['binding:host_id'] == target_host) and
```
Comment (aspiers) [inline files/default/neutron-ha-tool.py]: Same nitpick about parentheses here; `and` is lower precedence than `==`.

**Review by aspiers — CHANGES_REQUESTED** [review]
Basically looks great, just a few nitpicks.  Please tell me next time you start working on the thing I said I was hoping to work on though ;-) Luckily on this occasion there wasn't much duplicated effort :-)

File: files/default/neutron-ha-tool.py (hunk 2 above)
Comment (rhafer) [reply files/default/neutron-ha-tool.py]: Yeah, I will most likely change that. Lastest for the other feature we need to add (ssh support to remove the namespaces).

**Comment by rhafer** [comment]
> Basically looks great, just a few nitpicks. Please tell me next time you start working on the thing I said I was hoping to work on though ;-) Luckily on this occasion there wasn't much duplicated effort :-)

Actually I told you. Though I must admit I didn't wait too long for an answer. Maybe you just missed it.

File: files/default/neutron-ha-tool.py (hunk 3 above)
Comment (rhafer) [reply files/default/neutron-ha-tool.py]: I still need to look into that. Hence the "wip" label.

File: files/default/neutron-ha-tool.py (hunk 4 above)
Comment (rhafer) [reply files/default/neutron-ha-tool.py]: http://developer.openstack.org/api-ref/networking/v2/?expanded=list-ports-detail
https://wiki.openstack.org/wiki/Neutron/APIv2-specification#Filtering_and_Column_Selection

and the neutronclient source code :/

And yes the SDK API doc were not really helpful.

File: files/default/neutron-ha-tool.py (hunk 6 above)
Comment (rhafer) [reply files/default/neutron-ha-tool.py]: Hm, it's possible to get all router ports (and their details) in a single API call. Not sure which way is more efficient. But I guess you're right, getting them all at once will save us a couple of round trips to the server. I'll take a look at that.

File: files/default/neutron-ha-tool.py (hunk 7 above)
Comment (rhafer) [reply files/default/neutron-ha-tool.py]: Ack

File: files/default/neutron-ha-tool.py (hunk 1 above)
Comment (rhafer) [reply files/default/neutron-ha-tool.py]: Hm, I not really sure yet what's the best default here. So I kept backwards compatible. 

Does it really make sense in the HA fail-over use case (when a single node, that is most likely offline already, needs to be evacuated quickly) to add all the waiting? I'd think that in that case it's preferable to have this script finish as fast as possible.  And after all this is still the default use-case of neutron-ha-tool.

File: files/default/neutron-ha-tool.py [hunk 9]
```diff
@@ -339,34 +346,34 @@ def l3_agent_check(qclient):
...
             except IndexError:
                 LOG.warn("There are no l3 agents alive we could "
                          "migrate routers onto.")
-                target_id = None
+                target = None
 
             migration_count += 1
             LOG.warn("Would like to migrate router=%s to agent=%s",
-                     router_id, target_id)
+                     router_id, target['id'])
```
Comment (Itxaka) [inline files/default/neutron-ha-tool.py:377]: If `target` is `None` as set on line 367 this should raise a TypeError.

```
>>> import logging
>>> target = None
>>>logging.warn("Would like to migrate router=%s to agent=%s" % ("test", target['id']))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: 'NoneType' object has no attribute '__getitem__'
```

File: files/default/neutron-ha-tool.py [hunk 10]
```diff
@@ -516,39 +523,39 @@ def replicate_dhcp(qclient, noop=False):
...
         LOG.info("Would try to migrate router=%s from agent=%s "
-                 "to agent=%s", router_id, agent_id, target_id)
+                 "to agent=%s", router_id, agent['id'], target['id'])
         return True
 
     try:
-        migrate_router(qclient, router_id, agent_id, target_id)
+        migrate_router(qclient, router_id, agent, target, wait_for_ports)
         return True
     except:
```
Comment (Itxaka) [inline files/default/neutron-ha-tool.py:560]: Not part of this PR, but this is too broad and we are the ones raising (some of) the exceptions so we should try to catch the 2 know raises that we emit and then fall back to catching everything.

File: files/default/neutron-ha-tool.py [hunk 11]
```diff
@@ -559,31 +566,77 @@ def migrate_router(qclient, router_id, agent_id, target_id):
...
+                        a port)
+    :param maxtries: Maximum number of times the ports' status is polled,
+                     an expection is raised there are still ports with a
+                     non ACTIVE status after that.
+    """
+
+    # FIXME: Don't check the "network:router_interface_distributed" ports
+    # as those are not hosted on the network node
+    router_port_list = qclient.list_ports(device_id=router_id, fields='id')
+    port_ids = [r['id'] for r in router_port_list['ports']]
```
Comment (Itxaka) [inline files/default/neutron-ha-tool.py]: consider using `port` instead of `r`, this code is read by humans ;)

**Review by Itxaka — COMMENTED** [review]
Only comments codewise, I have no idea if the code actually works :D

File: files/default/neutron-ha-tool.py (hunk 9 above)
Comment (rhafer) [reply files/default/neutron-ha-tool.py:377]: Should be fixed now.

File: files/default/neutron-ha-tool.py (hunk 11 above)
Comment (rhafer) [reply files/default/neutron-ha-tool.py]: It's gone now. I reworked the code a bit while addressing @aspiers feedback.

File: files/default/neutron-ha-tool.py (hunk 5 above)
Comment (rhafer) [reply files/default/neutron-ha-tool.py]: Fixed

File: files/default/neutron-ha-tool.py (hunk 6 above)
Comment (rhafer) [reply files/default/neutron-ha-tool.py]: I changed it to use list_port instead of fetching the single ports one by one. It does not have any measurable effect on migration time. But I guess just keep it that way now .

File: files/default/neutron-ha-tool.py (hunk 8 above)
Comment (rhafer) [reply files/default/neutron-ha-tool.py]: Fixed

File: files/default/neutron-ha-tool.py (hunk 7 above)
Comment (rhafer) [reply files/default/neutron-ha-tool.py]: Fixed

File: files/default/neutron-ha-tool.py [hunk 12]
```diff
@@ -559,31 +566,81 @@ def migrate_router(qclient, router_id, agent_id, target_id):
...
+                        (reflected by the binding:host_id attribute of
+                        a port)
+    :param maxtries: Maximum number of times the ports' status is polled,
+                     an expection is raised there are still ports with a
+                     non ACTIVE status after that.
+    """
+
+    active_port_ids = set()
+    remaining = ['dummy']
+    LOG.info("Wait for the ports of router_id=%s to be ACTIVE" % router_id)
```
Comment (aplanas) [inline files/default/neutron-ha-tool.py]: nit: do not interpolate in logs: `LOG.info("Wait for the ports of router_id=%s to be ACTIVE", router_id)`

File: files/default/neutron-ha-tool.py [hunk 13]
```diff
@@ -559,31 +566,81 @@ def migrate_router(qclient, router_id, agent_id, target_id):
...
+                       port['status']))
+            if (port['binding:host_id'] == target_host and
+                port['status'] == 'ACTIVE'):
+                LOG.info("port_id=%s is ACTIVE on host_id=%s" %
+                         (port['id'], target_host))
+                active_port_ids.add(port['id'])
+            else:
+                remaining.append(port['id'])
+
+        if len(remaining):
```
Comment (aplanas) [inline files/default/neutron-ha-tool.py]: `if remaining`

File: files/default/neutron-ha-tool.py [hunk 14]
```diff
@@ -559,31 +566,81 @@ def migrate_router(qclient, router_id, agent_id, target_id):
...
+            time.sleep(1)
+            maxtries-=1
+
+    if len(remaining):
```
Comment (aplanas) [inline files/default/neutron-ha-tool.py]: same

File: files/default/neutron-ha-tool.py (hunk 2 above)
Comment (aplanas) [reply files/default/neutron-ha-tool.py]: IMHO is better to be explicit in the arguments passed, as is now.

File: files/default/neutron-ha-tool.py [hunk 15]
```diff
@@ -585,8 +589,58 @@ def migrate_router(qclient, router_id, agent, target):
...
+                        a port)
+    :param maxtries: Maximum number of times the ports' status is polled,
+                     an expection is raised there are still ports with a
+                     non ACTIVE status after that.
+    """
+
+    active_port_ids = set()
+    remaining = ['dummy']
+    LOG.info("Wait for the ports of router_id=%s to be ACTIVE" % router_id)
+    while (maxtries > 0 and len(remaining)):
```
Comment (aplanas) [inline files/default/neutron-ha-tool.py]: `while remaining and maxtries:` can do the trick, also I was thinking how to populate `remaining` before, and move the `maxtries` check inside the while and ends it with a `break`

File: files/default/neutron-ha-tool.py [hunk 16]
```diff
@@ -585,8 +589,58 @@ def migrate_router(qclient, router_id, agent, target):
...
+            LOG.debug("Router port=%s on hostid=%s in status=%s" %
+                      (port['id'], port['binding:host_id'],
+                       port['status']))
+            if (port['binding:host_id'] == target_host and
+                port['status'] == 'ACTIVE'):
+                LOG.info("port_id=%s is ACTIVE on host_id=%s" %
+                         (port['id'], target_host))
+                active_port_ids.add(port['id'])
+            else:
+                remaining.append(port['id'])
```
Comment (aplanas) [inline files/default/neutron-ha-tool.py]: IMHO this loop can be considered simplified with `set` operations. Basically we populate `remaining` (as a set) with `router_port_list['ports']` at the beginning, and we substract the set generated with the next `router_port_list['ports']` filtered with the condition on lines 628 and 629

File: files/default/neutron-ha-tool.py (hunk 12 above)
Comment (rhafer) [reply files/default/neutron-ha-tool.py]: @aplanas Hm, since I am not that deep into python, could you share some more details here on why it's wrong?

File: files/default/neutron-ha-tool.py (hunk 2 above)
Comment (rhafer) [reply files/default/neutron-ha-tool.py]: Up to a certain number of arguments you're right I guess. But with the next change we will probably hit a limit.

File: files/default/neutron-ha-tool.py (hunk 12 above)
Comment (Itxaka) [reply files/default/neutron-ha-tool.py]: I think that one of the big issues is that if there is an encoding error, it wont raise an exception and break the code but instead it will be logged, as the interpolation is not done on the code but on the logger part, which can deal with that.

File: files/default/neutron-ha-tool.py (hunk 16 above)
Comment (rhafer) [reply files/default/neutron-ha-tool.py]: ```router_port_list['ports']``` is a list of dicts. You can't put dicts into a ```set``` (a dict is not hashable)

File: files/default/neutron-ha-tool.py (hunk 12 above)
Comment (aplanas) [reply files/default/neutron-ha-tool.py]: Not wrong at all, just not expected:

https://github.com/rhafer/cookbook-openstack-network/blob/529cb51e47f65060cc51dd936c2b8f6daff263ca/files/default/neutron-ha-tool.py#L228

You can read the documentation about this here:
https://docs.python.org/2/library/logging.html#logging.debug

File: files/default/neutron-ha-tool.py (hunk 16 above)
Comment (aplanas) [reply files/default/neutron-ha-tool.py]: You are right, you need some data mangling to extract the new set that can be subtracted. I was referring to something like (not tested!!):

```python
remaining = {port['id'] for port in router_port_list['ports']}

while remaining:
    if not maxtries:
        break

    active_ports = {port['id'] for port in router_port_list['ports']
                    if port['binding:host_id'] == target_host and
                       port['status'] == 'ACTIVE')}
    remaining = remaining - active_ports
```

Comment (rhafer) [reply files/default/neutron-ha-tool.py]: It's not that easy, actually. The list of ports needs to be refetched with every iteration, as the port state will change. And they might even change in number (e.g. users might have added/delete ports meanwhile).
Additionally (e.g. for logging/debugging purposes) I'd like to keep the state for which ports are already completed. While avoiding to log them over and over again. If I adjust your above suggestion to include those two things I think the resulting code has **not** much of an advantage over  the current code.

Edited: Added the missing "not".

Comment (aplanas) [reply files/default/neutron-ha-tool.py]: OK for me

File: files/default/neutron-ha-tool.py (hunk 12 above)
Comment (rhafer) [reply files/default/neutron-ha-tool.py]: Thanks. And fixed.

File: files/default/neutron-ha-tool.py (hunk 13 above)
Comment (rhafer) [reply files/default/neutron-ha-tool.py]: Fixed

File: files/default/neutron-ha-tool.py (hunk 14 above)
Comment (rhafer) [reply files/default/neutron-ha-tool.py]: Fixed

File: files/default/neutron-ha-tool.py (hunk 16 above)
Comment (AbelNavarro) [reply files/default/neutron-ha-tool.py]: i'm with @aplanas here. It would be better to fetch that list of ports just once and then keep the active/remaining for the loop. Notice that this is a wait for ports to be active after the migration and we were living without it for long. So if we skip a port that was recently added/modified by the user we shouldn't care that much.

Comment (rhafer) [reply files/default/neutron-ha-tool.py]: @abel-navarro But we need to fetch the port list continuously otherwise, we will not get the state changes? The initial implementation I had pollled the ports one by one (using port_show), but that will cause a lot more round trips to the neutron server than just fetching all router ports. (See @aspiers comment and my replies on that: https://github.com/SUSE-Cloud/cookbook-openstack-network/pull/7#discussion_r91616666)

Comment (AbelNavarro) [reply files/default/neutron-ha-tool.py]: @rhafer I'm ok with the current implementation provide it has some significant advantage (maybe performance) over iterating on single ports. Otherwise I'd go to:

1. retrieve list of ports
2. iterate over ports that are not ACTIVE (polling single port)

Comment (rhafer) [reply files/default/neutron-ha-tool.py]: @abel-navarro on pretty small setup (2 network nodes, 50 routers with 2 ports each, no floating IPs) I don't see any significant difference (perfomance-wise) between either approaches.
But I could image that in a setup with many ports to poll that the aproach with  polling singles ports is less performant (if just because of the many Rest-API roundtrips that are required for that)

File: files/default/neutron-ha-tool.py (hunk 1 above)
Comment (aspiers) [reply files/default/neutron-ha-tool.py]: It's more important that the script has the desired effect than that it finishes ASAP.  And we already know that in production clouds, allowing a backlog to build in the OVS agents and/or neutron servers can cause the whole thing to fall over in a spectacular heap.  Correctness should trump performance.

File: files/default/neutron-ha-tool.py (hunk 2 above)
Comment (aspiers) [reply files/default/neutron-ha-tool.py]: Right.

File: files/default/neutron-ha-tool.py (hunk 3 above)
Comment (aspiers) [reply files/default/neutron-ha-tool.py]: OK thanks.

File: files/default/neutron-ha-tool.py (hunk 16 above)
Comment (aspiers) [reply files/default/neutron-ha-tool.py]: The customer cloud in question has many ports on most of the routers I looked at.

**Comment by aspiers** [comment]
Still think it should default to waiting, but willing to consider arguments to the contrary.

> Actually I told you. Though I must admit I didn't wait too long for an answer. Maybe you just missed it.

I was half-expecting you to say that, since as a rule it is much more likely that I missed something than that you communicated poorly ;-) In this case I now see a message in my IRC logs which I definitely missed at the time...

File: files/default/neutron-ha-tool.py (hunk 2 above)
Comment (rhafer) [reply files/default/neutron-ha-tool.py]: @aspiers Are you ok if I address this with https://github.com/SUSE-Cloud/cookbook-openstack-network/pull/8 ?

File: files/default/neutron-ha-tool.py (hunk 1 above)
Comment (rhafer) [reply files/default/neutron-ha-tool.py]: @aspiers I am still torn between the two possibilities here. You almost convinced me :).

I am just wondering if, by switching the default to ```True``` we make the release of this as an update to Cloud 6 more complicated than needed. AFAICS, if we don't switch the default we can just release this as is as an update to Cloud 6 (it shouldn't cause any difference in the behaviour, but provide what we need for the upgrade to work). And if we want to enable it for HA we'd just add an update to the ha barclamp that enables it  (maybe even providing a proposal attribute to switch) and reconfigures the pacemaker resource accordingly.

OTOH if we switch the default it seems to me (maybe I am wrong here) we would need to ship an update to the barclamp together with it. AFAIU when this feature is enable we need some config changes (timeouts at least) on the pacemaker resource.

So wouldn't we make our lives a bit easier by keeping the default?

File: files/default/neutron-ha-tool.py (hunk 16 above)
Comment (rhafer) [reply files/default/neutron-ha-tool.py]: Ok. I suggest to keep it to the current implementation for now.

File: files/default/neutron-ha-tool.py (hunk 2 above)
Comment (aspiers) [reply files/default/neutron-ha-tool.py]: @rhafer Sure.

File: files/default/neutron-ha-tool.py (hunk 1 above)
Comment (aspiers) [reply files/default/neutron-ha-tool.py]: > OTOH if we switch the default it seems to me (maybe I am wrong here) we would need to ship an update to the barclamp together with it. AFAIU when this feature is enable we need some config changes (timeouts at least) on the pacemaker resource.

@rhafer The `op start timeout` is already way too low as per [bsc#1011655 (neutron-ha-tool 120s start timeout too low for large clouds)](https://bugzilla.suse.com/show_bug.cgi?id=1011655), so we need an update to `crowbar-ha` anyway.

Comment (rhafer) [reply files/default/neutron-ha-tool.py]: @aspiers Ok. I changed the default. With that I believe I addressed all the concerns you raised.

**Comment by rhafer** [comment]
Hm, adding the WIP label back, because I still need to address the "FIXME" about DVR routers :/

**Review by aspiers — APPROVED** [review]
LGTM now - thanks!

**Comment by aspiers** [comment]
Oh yeah, I forgot about the DVR routers FIXME ...

File: files/default/neutron-ha-tool.py (hunk 10 above)
Comment (matelakat) [reply files/default/neutron-ha-tool.py:560]: FYI: Bare except is generally a bad thing to do https://docs.python.org/2/howto/doanddont.html#except

File: files/default/neutron-ha-tool.py [hunk 17]
```diff
@@ -415,10 +422,10 @@ def l3_agent_migrate(qclient, noop=False, now=False):
 
     total_errors = 0
     total_migrations = 0
-    for agent_id in agent_dead_list:
+    for agent in agent_dead_list:
         (migrations, errors) = \
             migrate_l3_routers_from_agent(
-                qclient, agent_id, agent_alive_list, noop)
+                qclient, agent, agent_alive_lqist, noop, wait_for_ports)
```
Comment (matelakat) [inline files/default/neutron-ha-tool.py]: You have a typo here, that will break things `s/agent_alive_lqist/agent_alive_list`

**Review by matelakat — CHANGES_REQUESTED** [review]
Please fix the typo

File: files/default/neutron-ha-tool.py (hunk 17 above)
Comment (rhafer) [reply files/default/neutron-ha-tool.py]: @matelakat Should be fixed now. Thanks for noticing

**Comment by rhafer** [comment]
I've reworked the loop in ```wait_router_ports_active``` once more. I think the currently implementation is somewhat cleaner (and more a long the line to what @aplanas suggested). It also should handle distributed routers correctly now, by not looking at port that have ```binding:vif_type=="distributed"```

**Comment by rhafer** [comment]
I noticed there is still one remaining bit. As floating IPs are not represented as ports on the router. We need to have a separate check for them to be migrated. I am currently working on that. Should have something ready soon.

**Comment by rhafer** [comment]
With the latest commit I think this change should be (feature) complete now. (So I removed the WIP label). Please review again. If preferred I can also squash the last three commit into one.

**Comment by matelakat** [comment]
@rhafer looking at it - rebasing my branch on your work.

File: files/default/neutron-ha-tool.py [hunk 18]
```diff
@@ -559,29 +574,88 @@ def migrate_router(qclient, router_id, agent_id, target_id):
...
+    :param maxtries: Maximum number of times the ports' status is polled,
+                     an expection is raised there are still ports with a
+                     non ACTIVE status after that.
+    """
+
+    LOG.info("Wait for the ports and floating IPs of router_id=%s "
+             "to be ACTIVE", router_id)
+    remaining_ports = ["dummy"]
+    remaining_fips = ["dummy"]
+    while maxtries:
```
Comment (matelakat) [inline files/default/neutron-ha-tool.py]: this block will always add a one second sleep, no matter how fast my ports become active, as initially both `remaining_fips` and `remaining_ports` are true. If I understand it correctly, the intention is to first check the ports, and then the floating ips, right? The other problem, is that the minimal value for `maxtries` should be 2, because one is taken by checking the ports, the other one is taken by checking the ips.

Let's use `continue` if we see that `remaining_ports` has emptied!

Also, could we add a separate variable instead of mutating `maxtries`, it is quite weird to read `maxtries -= 1`, something like this would read better:

```python
remaining_iterations = maxtries
while remaining_iterations:
    ...
    remaining_iterations -= 1
```

File: files/default/neutron-ha-tool.py [hunk 19]
```diff
@@ -559,29 +574,88 @@ def migrate_router(qclient, router_id, agent_id, target_id):
...
+                                                          'binding:host_id',
+                                                          'binding:vif_type'])
+            remaining_ports = [
+                port['id'] for port in router_port_list['ports']
+                if (port['binding:vif_type'] != 'distributed' and
+                    not (port['status'] == 'ACTIVE' and
+                         port['binding:host_id'] == target_host))
+            ]
+            LOG.debug("Ports not ACTIVE on router_id=%s: [%s]",
+                      router_id, ", ".join(remaining_ports))
```
Comment (matelakat) [inline files/default/neutron-ha-tool.py:638]: add:
```python
if not remaining_ports:
    continue
```

File: files/default/neutron-ha-tool.py [hunk 20]
```diff
@@ -559,29 +574,88 @@ def migrate_router(qclient, router_id, agent_id, target_id):
...
+        elif remaining_fips:
+            floating_ips = qclient.list_floatingips(router_id=router_id)
+            remaining_fips = [fip['id'] for fip in floating_ips['floatingips']
+                              if fip['status'] != 'ACTIVE']
+            LOG.debug("Floating IPs not active: [%s]",
+                      ", ".join(remaining_fips))
+
```
Comment (matelakat) [inline files/default/neutron-ha-tool.py:651]: add:
```python
if not remaining_fips:
    break
```

File: files/default/neutron-ha-tool.py [hunk 21]
```diff
@@ -559,29 +574,88 @@ def migrate_router(qclient, router_id, agent_id, target_id):
...
+        if remaining_ports or remaining_fips:
```
Comment (matelakat) [inline files/default/neutron-ha-tool.py]: We will no longer need this check I believe, we only need the sleep if the iteration is still non-zero.

**Review by matelakat — CHANGES_REQUESTED** [review]
Could you please amend the belly of `wait_router_migrated` - so it doesn't do a delay every time?

File: files/default/neutron-ha-tool.py (hunk 18 above)
Comment (rhafer) [reply files/default/neutron-ha-tool.py]: > this block will always add a one second sleep, no matter how fast my ports become active, as initially both remaining_fips and remaining_ports are true.

Hm, true. Thanks for noticing that.

>  If I understand it correctly, the intention is to first check the ports, and then the floating ips, right? 

Yeah right, it doesn't make much sense to start checking on the floating IPs before the ports a back.

**Comment by rhafer** [comment]
> Could you please amend the belly of wait_router_migrated - so it doesn't do a delay every time?

@matelakat I believe I addressed all the issues you raised.

**Review by matelakat — APPROVED** [review]

**Comment by rhafer** [comment]
@aspiers are you still ok with this?

**Comment by rhafer** [comment]
[bnc#1012968 [upgrade] router migration took too long during pre-upgrade of one node](https://trello.com/c/v0VZ3WLc/193-bnc-1012968-upgrade-router-migration-took-too-long-during-pre-upgrade-of-one-node)

**Review by aspiers — APPROVED** [review]
LGTM although I would prefer the squash.

**Comment by rhafer** [comment]
> LGTM although I would prefer the squash.

Squashed it. And adjusted the commit message.

**Comment by matelakat** [comment]
[P2: bsc#1008055: L3: mass migration of neutron routers is very slow](https://trello.com/c/ni7N64ev/1-p2-bsc-1008055-l3-mass-migration-of-neutron-routers-is-very-slow)

**Comment by matelakat** [comment]
[P2: bsc#1008055: L3: mass migration of neutron routers is very slow](https://trello.com/c/ni7N64ev/1-p2-bsc-1008055-l3-mass-migration-of-neutron-routers-is-very-slow)
